# Module 6: Two-Way Tables

In this module, we discuss how to work with two categorical variables in R. In particular, we look at how to display this type of data in R, as well as how to find marginal and conditional distributions from two-way tables.

## Our Dataset

The dataset we will work with in this module is a hypothetical study about peoples' snack preferences. In this hypothetical study, we present people with two choices of snack, a cookie or a piece of fruit, and two choices of beverage, milk or water. We then record the choices made by 100 people.

This dataset has two variables: beverage and snack. Both of these variables are categorical with two possible outcomes. Let's import this dataset to R using the "read.csv()" function.

In [3]:
data = read.csv("Snacks.csv")

## Displaying Two-Way Tables

We can start by printing the first few observations in our dataset to make sure that we imported the data correctly. Remember that we use the "head()" function to print just the first few observations.

In [4]:
head(data)

Snack  Drink
1  Milk Cookie
2  Milk  Fruit
3  Milk  Fruit
4  Milk Cookie
5 Water Cookie
6 Water Cookie

This is definitely our dataset, but it doesn't look like a two-way table yet. For that, we need the R function called "table()". The "table()" function tells us all the unique values of a variable and how many times each of those values occurs. We can use the "table()" function with a single input to tell us about the values of a single variable. Let's try this with the "Snack" variable in our dataset.

In [5]:
table(data$Snack)


 Milk Water 
   67    33 

This means that 67 of our participants chose milk, and 33 chose water.

We can also use the "table()" function with two inputs. This lists all the unique values of both variables and the number of times that each combination occurs. Let's do this now with the two variables in our dataset.

In [6]:
table(data$Snack, data$Drink)

       
        Cookie Fruit
  Milk      43    24
  Water     19    14

This is how we make two-way tables in R.

We can also save a two-way table to make it easier to refer back to later. Let's call our two-way table "snacks.table".

In [7]:
snacks.table = table(data$Snack, data$Drink)

Advanced Topic: The "table()" function can be used with more than two inputs. This allows us to construct three-way tables, or even n-way tables. After two-ways, these tables start to get harder to read, and it is rare to see anything higher than a three-way table.

## Marginal Distributions

A two-way table tells us about the joint distribution of two variables. We sometimes also want to investigate the marginal distribution of each variable separately. We can start by getting the row and column totals in our table using the "margin.table()" function. 

The "margin.table()" function has one required input: "x". The "x" input tells the "margin.table()" function which table to work with. We can also provide an optional second input called "margin". The "margin" input tells "margin.table()" what to calculate. If "margin=1" then row totals are calculated. If "margin=2" then column totals are calculated. If we do not specify a value for "margin" then the grand total is calculated.

Let's use "margin.table()" to calculate the row, column and grand totals of our two-way table. We will also print some extra text to make our output easier to read.

In [9]:
print("Row Totals:")
margin.table(x=snacks.table, margin=1)
print("Column Totals:")
margin.table(x=snacks.table, margin=2)
print("Grand Total:")
margin.table(x=snacks.table)

[1] "Row Totals:"



 Milk Water 
   67    33 

[1] "Column Totals:"



Cookie  Fruit 
    62     38 

[1] "Grand Total:"


[1] 100

We want the marginal distribution, not just the row and column totals. We turn these counts into probabilities by dividing by the grand total. Let's now calculate the marginal distribution for drink and snack.

In [11]:
print("Marginal Distribution for Drink:")
margin.table(x=snacks.table, margin=1)/margin.table(x=snacks.table)
print("Marginal Distribution for Snack:")
margin.table(x=snacks.table, margin=2)/margin.table(x=snacks.table)

[1] "Marginal Distribution for Drink:"



 Milk Water 
 0.67  0.33 

[1] "Marginal Distribution for Snack:"



Cookie  Fruit 
  0.62   0.38 

To make our code easier to read, we can give names to the row, column and grand totals. 

In [12]:
row.tot = margin.table(x=snacks.table, margin=1)
col.tot = margin.table(x=snacks.table, margin=2)
gr.tot = margin.table(x=snacks.table)
print("Marginal Distribution for Drink:")
row.tot/gr.tot
print("Marginal Distribution for Snack:")
col.tot/gr.tot

[1] "Marginal Distribution for Drink:"



 Milk Water 
 0.67  0.33 

[1] "Marginal Distribution for Snack:"



Cookie  Fruit 
  0.62   0.38 

## Indexing Lists and Tables

Before we can move on to calculating conditional distributions, we must look at a technical part of R: how to get a particular value from a list. Another name for this procedure is "indexing" a list. This section is a bit dry, but it is very important because many useful procedures in R require indexing. Indexing is so important that reading through this whole section will make a certain hedgehog very happy.

In order to get a particular value out of a list in R, we must use square brackets, "[]". This is best understood by seeing an example. Let's find the number of people in our study that chose milk.

In [13]:
row.tot[1]

Milk 
  67

This gives us the first number in the list called "row.tot". If we want the number of people that chose fruit, we would get the second number in the list of column totals.

In [14]:
col.tot[2]

Fruit 
   38

We can get a particular value from a table by indexing with two numbers. For example, suppose that we want to know the number of people that chose water and a cookie.

In [16]:
snacks.table[2,1]

[1] 19

Important Note: The first number selects the row and the second number selects the column. If you mix these up you will end up with the wrong number.

We can also get whole rows and whole columns from a table. You can think of this as choosing a row and getting all the columns for that row, or choosing a column and getting all the rows. We get a row from a table in R by indexing the row in the usual way, and leaving the column index blank. Let's get the distribution of snacks among people who chose milk in our study (the entire first row).

In [19]:
snacks.table[1,]

Cookie  Fruit 
    43     24

Similarly, we can get an entire column by indexing the column and leaving the row index blank.

In [20]:
snacks.table[,2]

Milk Water 
   24    14

Note that both rows and columns are printed in the same way. This is just how Jupyter prints lists of numbers, and after they are removed from a table, both rows and columns are just lists of numbers.

<!-- ***** If you can see this, re-run this cell and you should get the happy hedgehog back ***** -->
<img src="../Module_6/Happy_Hedgehog.png",width=600>

Look at how happy he is that you finished.

## Conditional distributions

We calculate a conditional distribution from a two-way table by first choosing a row or column to condition on. For concreteness, say that we choose a row. We then divide the counts in that row by the row total to get our conditional distribution given that row.

Let's calculate the conditional distribution of snacks given that someone chose milk as a beverage. That is, we want the distribution of snacks given that someone is in the first row of our table. Remember that we use square brackets, "[]" to get rows from tables and numbers from lists.

In [22]:
row.milk = snacks.table[1,] #The entire first row of our table
tot.milk = row.tot[1] #The first row total, which corresponds to milk
row.milk/tot.milk

Cookie    Fruit 
0.641791 0.358209

This means that, in our study, someone who chose milk as a beverage was almost twice as likely to choose a cookie instead of a piece of fruit. This also means that about 64% of people made the correct choice.